pretyping/inductiveops.mli

```diff
@@ -192,6 +192,10 @@ val find_rectype     : env -> evar_map -> EConstr.types -> inductive_type
 val find_inductive   : env -> evar_map -> EConstr.types -> (inductive * EConstr.EInstance.t) * constr list
 val find_coinductive : env -> evar_map -> EConstr.types -> (inductive * EConstr.EInstance.t) * constr list
 
+(** [instantiate_constructor_params ind mind i params] instantiates the
+    type of the i-th constructor (counting from 1) with parameters [params] *)
+val instantiate_constructor_params : pconstructor -> Inductive.mind_specif -> constr list -> constr
```

Comment:
The documentation looks wrong.

Reply:
Hum... now fixed.

Reply:
Since the function is brand new, would it make sense to change the signature from `constr list` to `constr array`? As it stands, the execution starts from an array, converts it to a list, then manipulates the list as if it was an array (e.g., it counts the number of elements in the list). This is getting a bit dumb, especially since the inner code (`Vars.subst_of_rel_context_instance`) does not even need it to be a list in the first place, as far as I can tell.

Reply:
Both `Vars.subst_of_rel_context_instance` and `subst_of_rel_context_instance_list` are currently implemented working on a list. Are you suggesting they work on an array (this was postponed in [#14794](https://github.com/coq/coq/pull/14794#discussion_r700187813) and maybe it is time to change it).

Reply:
Don't touch to much this API, it's sensitive performancewise. Arrays are pretty bad when it comes to compositionality because you have to pay full O(n) for a cons. I think we should merge this PR first and then nitpick about the current issues of our APIs.

Reply:
Anything is ok to me, but if one or the other wants me to eventually save an `Array.to_list` by rewriting in another PR `subst_of_rel_context_instance` directly on arrays (w/o changing `subst_of_rel_context_instance_list` so that it is sure that it cannot damage the complexity of the latter), I can do it.

Reply:
> I think we should merge this PR first

My point is that we should use the signature that makes sense in the long run. The new code systematically performs `instantiate_constructor_params (Array.to_list params)`. So, the signature is the wrong one. It should be doing `instantiate_constructor_params params`, even if the first thing this new function does is call `Array.to_list` to accommodate the existing functions.

Reply:
But there is a more common consumer of `instantiate_constructor_params` which is `Inductiveops.get_constructor` and this one has a list.

Reply:
But that one is in the same file, so it is not impacted by the signature in `inductiveops.mli`. All the functions that go through the public interface have an array at hand.

Reply:
But then, since `subst_of_rel_context_instance` is anyway working on a list and that a translation from array to list has to be done anyway at some time, I would not consider bad that the translation is done in `nativenorm.ml` vs having an extra version of `instantiate_constructor_params` in `inductivelops.ml` just to avoid the conversion to be done in `nativenorml.ml`.

I'm not saying that the choice list/array currently in use here and there is optimal. As @ppedrot proposed, we can eventually think globally at where to better use array and where to better use list, but `subst_of_rel_context_instance` being on list, I'm not convinced that the needed conversion should be in `inductiveops.ml` (to the price of an extra indirection) rather than in `nativenorm.ml`.

But again, I reiterate my proposal to rewrite `instantiate_constructor_param` directly on arrays so that the conversion is not needed for sure. That would solve two issues at once!

Reply:
I believe this discussion is veering out of topic. It's simpler to merge the PR as-is. @silene if you want to tweak the API please open a distinct PR.